topdown/object.go

```diff
@@ -81,11 +83,60 @@ func builtinObjectGet(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Ter
 		return err
 	}
 
-	if ret := object.Get(operands[1]); ret != nil {
-		return iter(ret)
+	// if the get key is not an array, attempt to get the top level key for the operand value in the object
+	path, err := builtins.ArrayOperand(operands[1].Value, 2)
+	if err != nil {
+		if ret := object.Get(operands[1]); ret != nil {
+			return iter(ret)
+		}
+
+		return iter(operands[2])
+	}
+
+	// if the path is empty, then we skip selecting nested keys and return the default
+	if path.Len() == 0 {
+		return iter(operands[2])
+	}
+
+	var pathTerm *ast.Term
+
+	// index is used to track if the term being processed is the last key term in the path
+	var index int
+	err = path.Iter(func(term *ast.Term) error {
+		index++
+
+		// pathTerm will either be the return value, or the object to dig for the next iteration
+		pathTerm = object.Get(term)
+		if pathTerm == nil {
+			return fmt.Errorf("pathTerm not found")
+		}
+
+		// if processing the final element, then we exit with whatever the current pathTerm is
+		if index == path.Len() {
+			return nil
+		}
+
+		// to continue getting nested keys from the path, the current pathTerm must be another object.
+		// if the pathTerm is not an object, then by returning an error the default value will be
+		// returned to the caller
+		var ok bool
+		object, ok = pathTerm.Value.(ast.Object)
+		if !ok {
+			return fmt.Errorf("pathTerm was not object")
+		}
+
+		return nil
+	})
```

Comment:
This could be replaced by turning the array into an ast.Ref and then calling `object.Find(ref)`. That should do the same the thing. As a bonus, it'll deal with nested arrays, e.g:

```
object.get({"foo": [{"bar": 7}]}, ["foo", 0, "bar"], -1) == 7
```

Reply:
Thanks, I think this is much better. In b6344ec1 I've had a go at this. Let me know if you expected it to be something different.

I think I am I right in saying this but can you confirm that this ref & find approach isn't suitable/available for the WASM implementation?